File: py/MinimumCostForTickets.py
# https://leetcode.com/problems/minimum-cost-for-tickets/description/

from functools import lru_cache
from typing import List


class Solution:
    def mincostTickets(self, days: List[int], costs: List[int]) -> int:
        dayExists, minCostAtDay = [False] * 366, [0] * 366
        for day in days:
            dayExists[day] = True
        for i in range(1, 366):
            if not dayExists[i]:
                minCostAtDay[i] = minCostAtDay[i - 1]  # if this day doesn't exist
            else:
                minCostAtDay[i] = min(
                    minCostAtDay[i - 1] + costs[0],
                    minCostAtDay[max(i - 7, 0)] + costs[1],
                    minCostAtDay[max(i - 30, 0)] + costs[2],
                )  # try to buy 7 day ticket at (i - 7)th day or 30 day ticket at (i - 30)th day
        return minCostAtDay[-1]

    def mincostTickets1(self, days: List[int], costs: List[int]) -> int:
        daysSet = set(days)

        @lru_cache(None)
        def dfs(i) -> int:  # min cost at ith day, starting from day[0]
            if i < days[0]:
                return 0
            if i == days[0]:
                return min(costs)
            if i not in daysSet:
                return dfs(i - 1)
            return min(
                dfs(i - 1) + costs[0],
                dfs(max(0, i - 7)) + costs[1],
                dfs(max(0, i - 30)) + costs[2],
            )

        return dfs(days[-1])


s = Solution()
days = [1, 4, 6, 7, 8, 20]
costs = [7, 2, 15]
print(s.mincostTickets(days, costs))
print(s.mincostTickets1(days, costs))

File: src/DP/MinimumCostForTickets.java
package DP;

// https://leetcode.com/problems/minimum-cost-for-tickets/discuss/226670/Java-DP-Solution-with-explanation-O(n)
// Similar with coin change
public class MinimumCostForTickets {
    public int mincostTickets(int[] days, int[] costs) {
        boolean[] daysExist = new boolean[366];
        for (int day : days) {
            daysExist[day] = true;
        }
        int[] minCostAtDay = new int[366]; // minCostAtDay[0] = 0;
        for (int i = 1; i <= 365; i++) {
            if (!daysExist[i]) {
                minCostAtDay[i] = minCostAtDay[i - 1]; // if this day doesn't exist
            } else {
                int min = minCostAtDay[i - 1] + costs[0]; // try to buy 1 day ticket
                min = Math.min(min, minCostAtDay[Math.max(i - 7, 0)] + costs[1]); // try to buy 7 day ticket at (i - 7)th day
                min = Math.min(min, minCostAtDay[Math.max(i - 30, 0)] + costs[2]); // try to buy 30 day ticket at (i - 30)th day
                minCostAtDay[i] = min;
            }
        }
        return minCostAtDay[365];
    }

    public static void main(String args[]) {
        int[] days = new int[]{1,4,6,7,8,20};
        int[] costs = new int[]{2,7,15};
        MinimumCostForTickets m = new MinimumCostForTickets();
        System.out.println(m.mincostTickets(days, costs));
    }

}
